_posts/2020-11-23-how-fast-are-ractors.md

```diff
@@ -0,0 +1,133 @@
+date: 2020-11-04 09:00:00
+categories: ["ruby", "performance"]
+author: "noahgibbs"
+---
+
+Ruby 3x3 is coming in about a month. One of its new concurrency primitives is [Ractors](https://github.com/ruby/ruby/blob/master/doc/ractor.md), which used to be called "Guilds." (The other new concurrency primitive is [Autofibers](https://bugs.ruby-lang.org/issues/13618).)
+
+Ruby has a [Global VM Lock (GVL)](https://www.speedshop.co/2020/05/11/the-ruby-gvl-and-scaling.html), also called the Global Interpreter Lock (GIL), that prevents running Ruby code in more than one thread at once. So Ruby threads are great for I/O like waiting on files or databases. And they're great for cases where a C extension can keep calculating in a background thread while a foreground thread runs Ruby. But you can't do calculations ***in Ruby*** in more than one thread at once within the same process.
+
+At least not until Ruby 3 and not without Ractors.
+
+Great! Now how fast is the current implementation of Ractors?
+
+<!--more-->
+
+## How Do Ractors Work?
+
+The idea goes like this: every Ractor in your process gets its own lock. And every Ractor can potentially hold multiple threads. So now threads hold their Ractor lock, not a single global lock. If you don't ever call Ractor.new then you'll still have one Global VM Lock just like as you always did. Or if you have six Ractors, you'll have six locks to share and you can fully use up to six cores of Ruby code all at once.
+
+Ractors have a clear idea of which Ractor holds what chunk of memory. They're almost like sub-VMs in your VM. You can pass objects between Ractors but it's fairly slow. There are objects that can be shared between Ractors, such as anything [frozen](https://www.rubyguides.com/2016/01/ruby-mutability/).
+
+[Here's example code for a multi-Ractor web server in Ruby](https://kirshatrov.com/2020/09/08/ruby-ractor-web-server/) by Kir Shatrov, if you'd like to see Ractors in action. But the code mostly looks a lot like Ruby processes or threads. The big difference is passing objects into and out of each Ractor.
+
+When and why might Ractors be better than existing Ruby threads? The lack of GVL allows them to fully use more cores. So in cases where Ruby can't easily use all your cores, Ractors can be better than threads.
+
+When and why might Ractors be better than multiprocess ("fork") concurrency in Ruby? They save more memory and have cheaper communication than multiple processes. So in a case with a lot of calculation and not a lot of ([working set](https://en.wikipedia.org/wiki/Working_set)) memory per worker, a Ractor can share memory efficiently and communicate quickly between workers. If this sounds a lot like threads in most languages (including JRuby) that's because it's essentially the same thing. CRuby's GVL kept it from efficiently threading Java-style. Ractors attempt to provide Java-style threading in Ruby, but with inter-thread communication more like Go or Erlang.
```

Comment:
> CRuby's GVL kept it from efficiently threading Java-style

@noahgibbs I'm curious here, why is that?

Reply:
Java has no GVL or equivalent. That means in Java you can use one core per thread for any kind of work. For Ruby, you only get one core total most of the time. Everything else has to be waiting on I/O (or, more rarely, doing some kind of C-based work that can release the GVL and doesn't touch Ruby objects.)

Reply:
It's common in Java to have a *lot* of threads efficiently - 10 or 20 more more, for instance. A Ruby process generally productively tops out at 5 to 6 for something I/O-heavy like a Rails app, and even fewer if it's primarily doing real calculation.

Reply:
Thinking on it, you're right. I shouldn't just leave that statement like that. I've put in a better (I think) explanation of Ruby threading compared to Java threading. Let me know if you'd like further changes.